xapian-applications/omega/postprocess.py

```diff
@@ -0,0 +1,182 @@
+        with open('log/final.log', 'w+') as h:
+            reader1 = csv.reader(f, delimiter='\t')
+            reader2 = csv.reader(g, delimiter='\t')
+            writer = csv.writer(h, delimiter='\t')
+
+            # Add headers to final log file
+            writer.writerow(["QueryID", "Query", "Hits", "Offset", "Clicks"])
+
+            qid_to_clicks = collections.defaultdict(dict)
+            did_to_count = {}
+
+            QUERYID, QUERY, HITS = 0, 1, 2
+            DOCID = 1
+
+            # Build map: qid_to_clicks = {qid: {did: click_count}}
+            for row2 in reader2:
+                qid, did = row2[QUERYID], row2[DOCID]
+
+                did_to_count = qid_to_clicks[qid]
+
+                # Check if did is already present in did_to_count
+                if did in did_to_count:
+                    # Update did_to_count[did]
+                    click_count = int(did_to_count[did])
+                    click_count += 1
+                    did_to_count[did] = str(click_count)
+                else:
+                    did_to_count[did] = '1'
+
+                qid_to_clicks[qid] = did_to_count
+
+            f.seek(0)
+
+            clicklist = []
+
+            for row in reader1:
+                # Skip rows with empty Query string
+                if row[QUERY] == '':
+                    continue
+
+                # Convert Hitlist from str to list
+                hits = row[HITS]
+                hits = hits.strip().split(',')
+                row[HITS] = hits
+
+                clicklist = hits[:]
+
+                # Update clicklist with click values stored in map.
+                if row[QUERYID] in qid_to_clicks:
+                    did_to_count = qid_to_clicks[row[QUERYID]]
+                    for index, did in enumerate(clicklist):
+                        if did in did_to_count:
+                            clicklist[index] = did + ':' + did_to_count[did]
+                        else:
+                            clicklist[index] = did + ':' + '0'
+
+                row.append(clicklist)
+                writer.writerow(row)
+    return
+
+
+def generate_query_file(final_log, query_file):
+    """Generates Query file formatted as per Xapian Letor documentation.
+
+    Input Args:
+        final_log (string): Path to final.log file.
+        query_file (string): Path to query.txt file.
+
+    Example (tab-delimited) entries in final.log:
+
+    QueryID	Query	Hits	Offset	Clicks
+    821f03288846297c2cf43c34766a38f7	book	['45', '36', '14', '54', '42',\
+    '52', '2', '3', '15', '32']	0	['45:0', '36:0', '14:0', '54:2', '42:0',\
+    '52:0', '2:0', '3:0', '15:0', '32:0']
+    098f6bcd4621d373cade4e832627b4f6	test	['45', '36', '14', '54', '42',\
+    '52', '2', '3', '15', '32']	0	['45:0', '36:0', '14:0', '54:0', '42:1',\
+    '52:0', '2:0', '3:0', '15:0', '32:0']
+
+    Example (comma-delimited) entries in query.txt:
+
+    821f03288846297c2cf43c34766a38f7,book
+    098f6bcd4621d373cade4e832627b4f6,test
```

Comment:
I'm not familiar with the internals of xapian letor but I think it's best to avoid duplicate entries here.